### PR TITLE
Basecamp job rework - wandering/sitting/job priorities.

### PR DESCRIPTION
**Pull request by ghost**
#### Summary

```SUMMARY: Features "Basecamp job rework - wandering/sitting/job priorities."```

#### Purpose of change
This is a combination of https://github.com/CleverRaven/Cataclysm-DDA/pull/37330 and https://github.com/CleverRaven/Cataclysm-DDA/pull/37325  , and a new feature of adding differing priorities to basecamp jobs.

Normally I wouldnt like to mush PRs together, but they were all chaining building on from one another, all caught up in freeze, rebasing wa sa nightmare, and as I worked on the 3rd one in the chain, I was discovering I had to rework the entire thing anyway which made the previous 2 PRs obsolete. So here it is.

Ive changed how jobs work - now it is its own class, not with an enum list of jobs, but an entry for each type of activity, which can then be given a priority each. This makes things more robust going forward with adding new tasks etc.

Ive changed how being assigned to a camp works, before it was an NPC mission, which conflicted with other stuff like the missoin for GUARD_ALLY and so on, it was best if being assigned to camp was a seperate thing that was more explicit, so NPCs didnt get confused about what to do.

Due to adding job priorities ive had to change how the activity code works a bit, so now they can just scan if an activity is possible, before they do it, if it isnt possible, they scan the next activity in their list.

Ive added them wandering and finding chairs to sit on when they dont have any tasks to do.

Also allowed activities to use the basecamp storage zone for the job components/fetching.

I dont *think* there will be any problems migrating from the previous job system in older saves, it should validate and the worst that would happen is that they dont do any jobs until you assign new priorities at the camp board, but im willing to accept ideas for how to migrate this fast moving mess a bit smoother.

This is something im laying some foundations for more flexible camps , where camp upgrades lay down construction zones, for this to work, it needs to be a bit smarter in how camp workers are assigned and do their jobs.

#### Describe the solution
As above

#### Describe alternatives you've considered
N/A
#### Testing
Made lots of new camps, assigned workers, killed them, checked if they were removed, told them to leave, made them mutiny, assigned jobs, assigned different priorities, layed down various types of tasks ( vehicle deconstruction and construction zones for example ) , chjecked to see they were performed in order, checked to see if they wandered around when thewy werent working.

Checked to see if th enew camp mission for "assign workers" worked by only showing followers, made sure that everything was serialized and deserialized correctly.
#### Additional context
N/A

### CODE REVIEW

File: src/basecamp.cpp [hunk 1]
```diff
@@ -479,15 +479,34 @@ void basecamp::reset_camp_workers()
...
-    for( auto it2 = assigned_npcs.begin(); it2 != assigned_npcs.end(); ) {
-        auto ptr = *it2;
-        if( ptr->mission != NPC_MISSION_ASSIGNED_CAMP || ptr->global_omt_location() != omt_pos ||
-            ptr->has_companion_mission() ) {
-            it2 = assigned_npcs.erase( it2 );
-        } else {
-            ++it2;
+    for( npc_ptr guy : assigned_npcs ) {
+        if( !guy || !guy->assigned_camp || *guy->assigned_camp != omt_pos ) {
+            assigned_npcs.erase( std::remove( assigned_npcs.begin(), assigned_npcs.end(), guy ),
```
Comment (BevapDin) [inline src/basecamp.cpp]: You're changing the very same container you're iterating over, that triggers UB. Consider iterating over the container via an iterator and use that iterator to remove the entry.

Comment (davidpwbrown) [reply src/basecamp.cpp]: OK now?

**Comment by Funguss** [comment]
So will job priorities be NPC specific? i.e. can I make Helen prioritise vehicle maintenance while Larry does some cooking then goes off to hunt?

**Comment by davidpwbrown** [comment]
> 
> 
> So will job priorities be NPC specific? i.e. can I make Helen prioritise vehicle maintenance while Larry does some cooking then goes off to hunt?

Yes. Each NPC will have their list of jobs with each having its own priority, like Rimworld.

**Comment by davidpwbrown** [comment]
Rebased, cleaned up, little bug fixed, tested, working.

**Comment by Qrox** [comment]
This is causing debug message to show when loading saves from 0.D experimental

```
 DEBUG    : Failed to read from "./save/Landrum/o.0.-1": line 31:7736: tried to start object, but found '0', not '{'

_coords":[233,-3],"last_player_seen_pos":null,"goalx":-2147483648,"goaly":-2147483648,"goalz":-2147483648,"guardx":-2147483648,"guardy":-2147483648,"guardz":-2147483648,"current_activity_id":"ACT_NULL","pulp_location":null,"mission":3,"job":0
                                                                                                                                                                                                                                                ^
                                                                                                                                                                                                                                                 ,"previous_mission":0,"faction_api_ver":2,"my_fac":"no_faction","attitude":0,"previous_attitude":0,"op_of_u":{"trust":0,"fear":0,"value":0,"anger":0,"owed":0},"chatbin":{"first_topic":"TALK_BUNKER_MERCHANT","skill":"none","style":"","missio


 FUNCTION : bool read_from_file(const string&, const std::function<void(std::basic_istream<char>&)>&)
 FILE     : *snip*/src/cata_utility.cpp
 LINE     : 434
```

Seems that "job" was changed from an int to an object but the deserialization code wasn't made back-compatible.

**Pull request merged**